### PR TITLE
Commit Position - Stateless Suggestion

### PR DESCRIPTION
**Pull request by nascimentolwtn**
Without storing totalCommits and hashes information. I liked this way in case of SubversionRepository.

Both branches has the same origin. I should have created a branch from your master branch to avoid other conflicts, my mistake.

Here is the result running the same study/visitor with this branch.

`
01:25:59  INFO Commit #d809ca97265620eb41f1ec87fa74ccfc9049b4cf @org.ops4j.pax.cdi (100.00%) in 18/03/2016 06:20:34 from Antonin Stefanutti with 1 modifications
01:25:59  INFO -> Processing d809ca97265620eb41f1ec87fa74ccfc9049b4cf with apache-evolution_org.ops4j.pax.cdi
01:25:59  INFO # --------------------------------------------------
01:25:59  INFO Study has been executed in the following projects:
01:25:59  INFO - https://github.com/ops4j/org.ops4j.pax.cdi, from 0c0d43c245c3c3e1195532cb9e80ed44bdc8f506 to d809ca97265620eb41f1ec87fa74ccfc9049b4cf
01:25:59  INFO The following processors were executed:
01:25:59  INFO - apache-evolution_org.ops4j.pax.cdi (br.inpe.cap.apache.ApacheEvolutionVisitor)
01:25:59  INFO Finished: 11-03-2016 01:25:59
01:25:59  **INFO It took 35 seconds (~0 minutes).**
01:25:59  INFO Brought to you by RepoDriller (repodriller.org.br)
01:25:59  INFO # -------------------------------------------------- #
`

### CODE REVIEW

File: src/main/java/org/repodriller/scm/GitRepository.java [hunk 1]
```diff
@@ -232,7 +232,11 @@ public Commit getCommit(String id) {
 				if(jgitCommit.getParentCount() > 1) merge = true;
 				Set<String> branches = getBranches(git, hash);
 				boolean isCommitInMainBranch = branches.contains(this.mainBranchName);
-				theCommit = new Commit(hash, author, committer, authorDate, authorTimeZone, committerDate, committerTimeZone, msg, parent, merge, branches, isCommitInMainBranch);
+				List<ChangeSet> changeSets = getChangeSets();
+				int totalCommits = changeSets.size();
+				long commitPosition = totalCommits - changeSets.stream().map((cs)->cs.getId()).collect(Collectors.toList()).indexOf(hash);
```
Comment (nascimentolwtn) [inline src/main/java/org/repodriller/scm/GitRepository.java]: Although this command is written in 1 line, it maps hashes through all commits every getCommit()

**Comment by mauricioaniche** [comment]
I'd love to *not* put this logic into GitRepository or RepositoryMining. These classes are already big enough and every linter complains about them! :/

Let me think a bit!

**Comment by nascimentolwtn** [comment]
GitRepository and RepositoryMining are huge classes indeed. The change in
RepositoryMining is just another log.info to help follow the mining
process.

But in the case of GitRepository changes, it is the one that constructs
Commit, and commitPosition is a property of Commit and somehow should be
encapsulated, shouldn't it? Just like in SubversionRepository, that the
revisionNumber is a direct info about SVNLogEntry.

What if we create some helpers to GitRepository?

**Comment by mauricioaniche** [comment]
I was thinking more towards creating an utility class like `CommitCounter`. If the visitor implements `RequiresCommitCounter` then we inject this class there. This class then does the magic.

What do you think?

**Comment by nascimentolwtn** [comment]
It's nice to 'componentize' the design. `CommitCounter` would be visitor's component, then something like `((RequiresCommitCounter) visitor).count(commit)` would be called from `RepositoryMine.processChangeSet` to provide commitPosition and % to it's visitor and it could also be logged.info from there?

**Comment by mauricioaniche** [comment]
My idea is as follows: if the Visitor also implements `RequiresCommitCounter`, it has to implement a `setCommitCounter(CommitCounter cc)`.

RepoDriller should inject this `CommitCounter` if he notices the visitor requires a commit counter.  `CommitCounter` does the magic. I suggest a method `count` that receives the commit and returns a `CommitPosition`. This class contains two fields: absolutePosition and relativePosition.

```
class MyVisitor implements CommitVisitor, RequiresCommitCounter {
  private CommitCounter cc;

  public void setCommitCounter(CommitCounter cc) {
    this.cc = cc;
  }

  public void process(Commit c, ...) {
    CommitPosition position = cc.count(commit);
  }
}
```

Question is where to make the injection. I think we have no choice rather than doing it in the `mine()` method.

What do you think?

**Comment by nascimentolwtn** [comment]
The `RequiresCommitCounter` goes fine with no issues, as you suggested. But the `CommitCounter` will need the `List<ChangeSets>` from `range.get(repo)`. It would be better injected and initialized with #19 start callback feature. Maybe I can try to implement that first.

**Comment by nascimentolwtn** [comment]
Using the visitor initialize method (#58), the `RequiresCommitCounter` interface will be something like this:
`
public interface RequiresCommitCounter {
	
	default void initialize(SCMRepository repo, PersistenceMechanism writer) {
		List<ChangeSet> changeSets = repo.getScm().getChangeSets();
		List<String> hashes = changeSets.stream().map((cs)->cs.getId()).collect(Collectors.toList());
		this.setCommitCounter(new CommitCounter(hashes));
	}
	void setCommitCounter(CommitCounter commitCounter);
	void count(Commit commit);
}
`

I still have to write some test code to validate this initalization override default method, and when we stabilize #58 I can merge this branch to that branch. What do you think?

**Comment by mauricioaniche** [comment]
We never continued here, @nascimentolwtn! Should I close this one?

**Comment by mauricioaniche** [comment]
Closing it for now!

**Pull request closed without merging**